File: packages/node/src/router.ts
/**
 * Main Router -- the primary public API for TryAii-DRE.
 *
 * Usage:
 *   import { Router } from 'tryaii-dre';
 *
 *   const router = new Router();
 *   const result = await router.route('Write a Python function to merge sorted arrays');
 *   console.log(result.bestModel);   // e.g., "gpt-5.2"
 *   console.log(result.scores);      // Top models with scores and reasoning
 *
 * The default `LocalEmbeddingProvider` is async-only, so `route()` itself is
 * async. Callers that have injected a sync embedding provider can use the
 * niche `routeSync()` method for a blocking call.
 */

import { BenchmarkRegistry, BenchmarkDefinition } from './benchmarks/registry.js';
import { NormalizationRange } from './scoring/benchmarks.js';
import { CentroidLoader } from './centroids/loader.js';
import { ClassificationResult } from './classifiers/base.js';
import { EmbeddingClassifier } from './classifiers/embedding.js';
import { TryaiiDreConfig, createDefaultConfig, centroidFilePath } from './config.js';
import { BaseEmbeddingProvider } from './embeddings/base.js';
import { LocalEmbeddingProvider } from './embeddings/local.js';
import { ModelRegistry, ModelInfo } from './registry/models.js';
import { ModelScore, ScoringEngine } from './scoring/engine.js';
import { DEFAULT_PRIORITIES, Priorities } from './scoring/priorities.js';
import type { LatencyTier } from './types.js';

/**
 * Result of routing a prompt.
 *
 * Contains the recommended model, all scored models, and classification details.
 */
export interface RouteResult {
  /** The top recommended model ID. */
  bestModel: string;

  /** All scored models (sorted by score descending). */
  scores: ModelScore[];

  /** Classification details (what kind of task was detected). */
  classification: ClassificationResult | null;

  /** Priorities used for this routing decision. */
  priorities: Priorities;
}

/** Get list of model IDs in ranked order from a RouteResult. */
export function routeResultTopK(result: RouteResult): string[] {
  return result.scores.map((s) => s.modelId);
}

/** Get the score of the top model. */
export function routeResultBestScore(result: RouteResult): number {
  return result.scores[0]?.finalScore ?? 0;
}

/** Get the reasoning for why the top model was chosen. */
export function routeResultBestReasoning(result: RouteResult): string {
  return result.scores[0]?.reasoning ?? '';
}

/** Options for route(). */
export interface RouteOptions {
  /** Quality/cost/speed priorities. Defaults to balanced. */
  priorities?: Priorities;
  /** Number of top models to return. */
  topK?: number;
  /** Only consider models from this provider. */
  filterProvider?: string;
  /** Only consider models with this capability. */
  filterCapability?: string;
  /** Only consider models cheaper than this (input $/1k tokens). */
  filterMaxCost?: number;
}

/**
 * Semantic AI model router.
 *
 * Analyzes user prompts using embeddings, matches them against benchmark
 * centroids, and recommends the best AI model based on benchmark performance,
 * pricing, latency, and user priorities.
 */
export class Router {
  private _config: TryaiiDreConfig;
  private _registry: ModelRegistry;
  private _benchmarkRegistry: BenchmarkRegistry;
  private _scoringEngine: ScoringEngine;
  private _embeddingProvider: BaseEmbeddingProvider | null;
  private _centroidLoader: CentroidLoader | null = null;
  private _classifier: EmbeddingClassifier | null = null;

  constructor(opts?: {
    config?: Partial<TryaiiDreConfig>;
    registry?: ModelRegistry;
    benchmarkRegistry?: BenchmarkRegistry;
    embeddingProvider?: BaseEmbeddingProvider;
  }) {
    this._config = createDefaultConfig(opts?.config);

    // Model registry
    this._registry = opts?.registry ?? ModelRegistry.default();

    // Benchmark registry
    this._benchmarkRegistry = opts?.benchmarkRegistry ?? BenchmarkRegistry.default();

    // Scoring engine with normalizer from benchmark registry
    const normalizer = this._benchmarkRegistry.getNormalizer();
    this._scoringEngine = new ScoringEngine(normalizer);

    // Embedding provider (lazy -- only initialized when needed)
    this._embeddingProvider = opts?.embeddingProvider ?? null;
  }

  /**
   * Lazy-initialize the embedding provider + centroid loader.
   *
   * The loader is stored on `this` so that `addBenchmark()` mutates the
   * same instance the classifier reads from -- custom benchmarks added
   * via the Router become visible to subsequent `route()` calls without
   * re-instantiating anything.
   */
  private _ensureCentroidLoader(): CentroidLoader {
    if (this._centroidLoader !== null) return this._centroidLoader;

    if (this._embeddingProvider === null) {
      this._embeddingProvider = new LocalEmbeddingProvider(
        `Xenova/${this._config.embeddingModel}`,
      );
    }

    this._centroidLoader = new CentroidLoader(
      this._embeddingProvider,
      centroidFilePath(this._config),
    );

    return this._centroidLoader;
  }

  /** Lazy-initialize the embedding classifier on first use. */
  private _ensureClassifier(): EmbeddingClassifier {
    if (this._classifier !== null) return this._classifier;

    const centroidLoader = this._ensureCentroidLoader();
    // `_embeddingProvider` is non-null after _ensureCentroidLoader returns.
    const provider = this._embeddingProvider as BaseEmbeddingProvider;

    this._classifier = new EmbeddingClassifier(
      provider,
      centroidLoader,
      {
        embeddingCacheSize: this._config.cache.embeddingCacheSize,
        classificationCacheSize: this._config.cache.classificationCacheSize,
        ttlSeconds: this._config.cache.ttlSeconds,
      },
    );

    return this._classifier;
  }

  /**
   * Route a prompt to the best AI model.
   *
   * Async by default -- works with any embedding provider, including the
   * default async `LocalEmbeddingProvider`. For a blocking call backed by a
   * sync provider, see `routeSync()`.
   *
   * @param prompt - The user's input text to classify and route.
   * @param opts - Routing options (priorities, filters, topK).
   * @returns RouteResult with the best model and full scoring breakdown.
   */
  async route(prompt: string, opts?: RouteOptions): Promise<RouteResult> {
    const classifier = this._ensureClassifier();
    const classification = await classifier.classifyAsync(prompt);
    return this._buildResult(classification, opts);
  }

  /**
   * Synchronous version of `route()`.
   *
   * Requires the injected embedding provider to support sync calls
   * (`supportsSync === true`). Throws otherwise -- the default
   * `LocalEmbeddingProvider` is async-only, so calling `routeSync()` on a
   * default `Router` will fail. Inject a sync provider (e.g. a custom
   * cached provider) to use this path.
   */
  routeSync(prompt: string, opts?: RouteOptions): RouteResult {
    const classifier = this._ensureClassifier();
    if (this._embeddingProvider !== null && !this._embeddingProvider.supportsSync) {
      throw new Error(
        `routeSync() requires an embedding provider that supports sync calls. ` +
          `${this._embeddingProvider.constructor.name} is async-only -- ` +
          `use route() (async) instead, or inject a sync provider.`,
      );
    }
    const classification = classifier.classify(prompt);
    return this._buildResult(classification, opts);
  }

  /** Shared post-classification path: filter models, score, return RouteResult. */
  private _buildResult(
    classification: ClassificationResult,
    opts?: RouteOptions,
  ): RouteResult {
    const priorities = opts?.priorities ?? DEFAULT_PRIORITIES;
    const topK = opts?.topK ?? 5;

    let models = this._registry.allModels;
    if (opts?.filterProvider) {
      const providerLower = opts.filterProvider.toLowerCase();
      models = models.filter((m) => m.provider.toLowerCase() === providerLower);
    }
    if (opts?.filterCapability) {
      const cap = opts.filterCapability;
      models = models.filter((m) => m.capabilities.includes(cap));
    }
    if (opts?.filterMaxCost != null) {
      const maxCost = opts.filterMaxCost;
      models = models.filter((m) => m.pricing != null && m.pricing.inputPer1k <= maxCost);
    }

    if (models.length === 0) {
      return {
        bestModel: '',
        scores: [],
        classification,
        priorities,
      };
    }

    const scores = this._scoringEngine.scoreModels(
      models,
      classification.benchmarkScores,
      priorities,
      topK,
    );

    const best = scores[0]?.modelId ?? '';

    return {
      bestModel: best,
      scores,
      classification,
      priorities,
    };
  }

  /**
   * Shortcut to add a model to the registry.
   *
   * @see ModelRegistry.add()
   */
  addModel(opts: {
    modelId: string;
    provider: string;
    benchmarks?: Record<string, number>;
    pricing?: [number, number];
    latency?: LatencyTier;
    capabilities?: string[];
    description?: string;
  }): ModelInfo {
    return this._registry.add(opts);
  }

  /**
   * Add a custom benchmark to the routing system.
   *
   * Async by default -- works with any embedding provider, including the
   * async-only `LocalEmbeddingProvider`. The new centroid is generated
   * immediately and added to the shared centroid loader, so subsequent
   * `route()` calls see it without any restart.
   *
   * For sync-provider callers who want a blocking setup step, use
   * `addBenchmarkSync()`.
   *
   * @param name - Benchmark name (e.g., "CustomerSupportQA").
   * @param queries - Representative prompts for this benchmark (10-20 recommended).
   * @param description - Human-readable description.
   * @param minScore - Minimum score for normalization.
   * @param maxScore - Maximum score for normalization.
   */
  async addBenchmark(
    name: string,
    queries: string[],
    description = '',
    minScore = 0,
    maxScore = 100,
  ): Promise<void> {
    this._registerBenchmark(name, queries, description, minScore, maxScore);
    const loader = this._ensureCentroidLoader();
    await loader.addBenchmarkCentroidAsync(name, queries);
  }

  /**
   * Synchronous version of `addBenchmark()`.
   *
   * Requires the injected embedding provider to support sync calls
   * (`supportsSync === true`). The default `LocalEmbeddingProvider` is
   * async-only, so calling `addBenchmarkSync()` on a default `Router`
   * will throw.
   */
  addBenchmarkSync(
    name: string,
    queries: string[],
    description = '',
    minScore = 0,
    maxScore = 100,
  ): void {
    this._registerBenchmark(name, queries, description, minScore, maxScore);
    const loader = this._ensureCentroidLoader();
    if (this._embeddingProvider !== null && !this._embeddingProvider.supportsSync) {
      throw new Error(
        `addBenchmarkSync() requires an embedding provider that supports sync calls. ` +
          `${this._embeddingProvider.constructor.name} is async-only -- ` +
          `use addBenchmark() (async) instead, or inject a sync provider.`,
      );
    }
    loader.addBenchmarkCentroid(name, queries);
  }

  /** Register a benchmark in the registry and rebuild the scoring normalizer. */
  private _registerBenchmark(
    name: string,
    queries: string[],
    description: string,
    minScore: number,
    maxScore: number,
  ): void {
    const benchmark: BenchmarkDefinition = {
      name,
      description,
      trainingQueries: queries,
      normalization: new NormalizationRange(minScore, maxScore, description),
      broadCategory: 'TECHNICAL',
      subcategories: [],
      metadata: {},
    };
    this._benchmarkRegistry.register(benchmark);

    const normalizer = this._benchmarkRegistry.getNormalizer();
    this._scoringEngine = new ScoringEngine(normalizer);
  }

  /** Access the model registry. */
  get models(): ModelRegistry {
    return this._registry;
  }

  /** Access the benchmark registry. */
  get benchmarks(): BenchmarkRegistry {
    return this._benchmarkRegistry;
  }

  /** Access the configuration. */
  get config(): TryaiiDreConfig {
    return this._config;
  }
}

File: packages/python/tryaii_dre/router.py
"""
Main Router -- the primary public API for TryAii-DRE.

Usage:
    from tryaii_dre import Router

    router = Router()
    result = router.route("Write a Python function to merge sorted arrays")
    print(result.best_model)     # "gpt-5.2"
    print(result.scores[:3])     # Top 3 with scores and reasoning
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Optional

from tryaii_dre.benchmarks.registry import BenchmarkRegistry
from tryaii_dre.centroids.loader import CentroidLoader
from tryaii_dre.classifiers.base import MAX_PROMPT_LENGTH, ClassificationResult
from tryaii_dre.classifiers.embedding import EmbeddingClassifier
from tryaii_dre.config import TryaiiDreConfig
from tryaii_dre.embeddings.local import LocalEmbeddingProvider
from tryaii_dre.registry.models import ModelRegistry
from tryaii_dre.scoring.engine import ModelScore, ScoringEngine
from tryaii_dre.scoring.priorities import DEFAULT_PRIORITIES, Priorities

logger = logging.getLogger("tryaii_dre")


@dataclass
class RouteResult:
    """
    Result of routing a prompt.

    Contains the recommended model, all scored models, and classification details.
    """

    # The top recommended model ID
    best_model: str

    # All scored models (sorted by score descending)
    scores: list[ModelScore] = field(default_factory=list)

    # Classification details (what kind of task was detected)
    classification: Optional[ClassificationResult] = None

    # Priorities used for this routing decision
    priorities: Priorities = field(default_factory=lambda: DEFAULT_PRIORITIES)

    @property
    def top_k(self) -> list[str]:
        """List of model IDs in ranked order."""
        return [s.model_id for s in self.scores]

    @property
    def best_score(self) -> float:
        """Score of the top model."""
        return self.scores[0].final_score if self.scores else 0.0

    @property
    def best_reasoning(self) -> str:
        """Reasoning for why the top model was chosen."""
        return self.scores[0].reasoning if self.scores else ""

    def __repr__(self) -> str:
        top3 = ", ".join(f"{s.model_id}({s.final_score:.2f})" for s in self.scores[:3])
        return f"RouteResult(best={self.best_model}, top3=[{top3}])"


class Router:
    """
    Semantic AI model router.

    Analyzes user prompts using embeddings, matches them against benchmark
    centroids, and recommends the best AI model based on benchmark performance,
    pricing, latency, and user priorities.

    Args:
        config: Configuration overrides. If None, uses defaults.
        registry: Model registry. If None, loads the default 35+ models.
        benchmark_registry: Benchmark definitions. If None, uses standard 12.
        embedding_provider: Custom embedding provider. If None, uses local
                           sentence-transformers (all-MiniLM-L6-v2).
    """

    def __init__(
        self,
        config: Optional[TryaiiDreConfig] = None,
        registry: Optional[ModelRegistry] = None,
        benchmark_registry: Optional[BenchmarkRegistry] = None,
        embedding_provider=None,
    ):
        self._config = config or TryaiiDreConfig()

        # Model registry
        self._registry = registry or ModelRegistry.default()

        # Benchmark registry
        self._benchmark_registry = benchmark_registry or BenchmarkRegistry.default()

        # Scoring engine with normalizer from benchmark registry
        normalizer = self._benchmark_registry.get_normalizer()
        self._scoring_engine = ScoringEngine(normalizer=normalizer)

        # Embedding provider (lazy -- only initialized when needed)
        self._embedding_provider = embedding_provider
        self._classifier: Optional[EmbeddingClassifier] = None

    def _ensure_classifier(self) -> EmbeddingClassifier:
        """Lazy-initialize the embedding classifier on first use."""
        if self._classifier is not None:
            return self._classifier

        # Initialize embedding provider
        if self._embedding_provider is None:
            self._embedding_provider = LocalEmbeddingProvider(
                model_name=self._config.embedding_model
            )

        # Initialize centroid loader
        centroid_loader = CentroidLoader(
            config=self._config,
            embedding_provider=self._embedding_provider,
        )

        self._classifier = EmbeddingClassifier(
            embedding_provider=self._embedding_provider,
            centroid_loader=centroid_loader,
            config=self._config,
        )

        return self._classifier

    def route(
        self,
        prompt: str,
        priorities: Optional[Priorities] = None,
        top_k: int = 5,
        filter_provider: Optional[str] = None,
        filter_capability: Optional[str] = None,
        filter_max_cost: Optional[float] = None,
    ) -> RouteResult:
        """
        Route a prompt to the best AI model.

        Args:
            prompt: The user's input text to classify and route.
            priorities: Quality/cost/speed priorities. Defaults to balanced.
            top_k: Number of top models to return.
            filter_provider: Only consider models from this provider.
            filter_capability: Only consider models with this capability.
            filter_max_cost: Only consider models cheaper than this (input $/1k tokens).

        Returns:
            RouteResult with the best model and full scoring breakdown.
        """
        if not prompt or not isinstance(prompt, str):
            raise ValueError("prompt must be a non-empty string")
        if len(prompt) > MAX_PROMPT_LENGTH:
            prompt = prompt[:MAX_PROMPT_LENGTH]

        priorities = priorities or DEFAULT_PRIORITIES

        # 1. Classify the prompt
        classifier = self._ensure_classifier()
        classification = classifier.classify(prompt)

        # 2. Get available models (with optional filters)
        models = self._registry.all_models
        if filter_provider:
            models = [m for m in models if m.provider.lower() == filter_provider.lower()]
        if filter_capability:
            models = [m for m in models if filter_capability in m.capabilities]
        if filter_max_cost is not None:
            models = [
                m for m in models
                if m.pricing and m.pricing.input_per_1k <= filter_max_cost
            ]

        if not models:
            return RouteResult(
                best_model="",
                scores=[],
                classification=classification,
                priorities=priorities,
            )

        # 3. Score and rank models
        scores = self._scoring_engine.score_models(
            models=models,
            benchmark_similarities=classification.benchmark_scores,
            priorities=priorities,
            top_k=top_k,
        )

        best = scores[0].model_id if scores else ""

        logger.info("Route completed best=%s category=%s confidence=%.3f top_k=%d",
                     best, classification.broad_category, classification.confidence, len(scores))

        return RouteResult(
            best_model=best,
            scores=scores,
            classification=classification,
            priorities=priorities,
        )

    def add_model(self, *args, **kwargs):
        """Shortcut to add a model to the registry. See ModelRegistry.add()."""
        return self._registry.add(*args, **kwargs)

    def add_benchmark(
        self,
        name: str,
        queries: list[str],
        description: str = "",
        min_score: float = 0,
        max_score: float = 100,
    ) -> None:
        """
        Add a custom benchmark to the routing system.

        Args:
            name: Benchmark name (e.g., "CustomerSupportQA").
            queries: Representative prompts for this benchmark (10-20 recommended).
            description: Human-readable description.
            min_score: Minimum score for normalization.
            max_score: Maximum score for normalization.
        """
        from tryaii_dre.benchmarks.registry import BenchmarkDefinition
        from tryaii_dre.scoring.benchmarks import NormalizationRange

        # Register in benchmark registry
        benchmark = BenchmarkDefinition(
            name=name,
            description=description,
            training_queries=queries,
            normalization=NormalizationRange(min_score, max_score, description),
        )
        self._benchmark_registry.register(benchmark)

        # Update scoring engine normalizer
        normalizer = self._benchmark_registry.get_normalizer()
        self._scoring_engine = ScoringEngine(normalizer=normalizer)

        # Generate centroid if classifier is initialized
        if self._classifier is not None and self._embedding_provider is not None:
            centroid_loader = CentroidLoader(
                config=self._config,
                embedding_provider=self._embedding_provider,
            )
            centroid_loader.add_benchmark_centroid(name, queries)

        logger.info(f"Added custom benchmark: {name} ({len(queries)} queries)")

    @property
    def models(self) -> ModelRegistry:
        """Access the model registry."""
        return self._registry

    @property
    def benchmarks(self) -> BenchmarkRegistry:
        """Access the benchmark registry."""
        return self._benchmark_registry

    @property
    def config(self) -> TryaiiDreConfig:
        """Access the configuration."""
        return self._config
